core/commands/pin.go

```diff
@@ -61,22 +68,84 @@ var addPinCmd = &cmds.Command{
 			res.SetError(err, cmds.ErrNormal)
 			return
 		}
+		showProgress, _, _ := req.Option("progress").Bool()
 
-		added, err := corerepo.Pin(n, req.Context(), req.Arguments(), recursive)
-		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
+		if !showProgress {
+			added, err := corerepo.Pin(n, req.Context(), req.Arguments(), recursive)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			res.SetOutput(&AddPinOutput{Pins: added})
 			return
 		}
 
-		res.SetOutput(&PinOutput{added})
+		v := new(dag.ProgressTracker)
+		ctx := context.WithValue(req.Context(), "progress", v)
+
+		ch := make(chan []*cid.Cid)
+		go func() {
+			defer close(ch)
+			added, err := corerepo.Pin(n, ctx, req.Arguments(), recursive)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			ch <- added
+		}()
+		out := make(chan interface{})
+		res.SetOutput((<-chan interface{})(out))
+		go func() {
+			defer close(out)
+			for {
+				timer := time.NewTimer(500 * time.Millisecond)
```

Comment:
Dear @whyrusleeping,

I did make a mistake, but the mistake I think you were pointing out:

> They build up and generate events until you close them. Do not make this mistake, it causes weird side effects

And calling stopTimer which I thought I did should removes them from the heap, https://golang.org/src/runtime/time.go#L71.

This mistake I made was assuming that the defer in a loop would fire before the next iteration.  This is how a similar construct would of worked in C++.  It seams I was wrong and go takes a more simple minded approach.  The defer does not fire until the end of the function as demonstrated in this code (https://play.golang.org/p/zKTz1Kc0K_)
```
func main() {
	for i := 0; i < 10; i++ { 
		defer fmt.Println("ByBy")
		println("Hello")
	}
}
```
I personally think go is wrong here, but that is not something I wish to have an extended discussion on.  I will just know to keep in mind how defer works in go and try not to make similar assumptions based on how C++ works, as go is obviously not C++.